### PR TITLE
WIP profiler / telnet fixes

### PR DESCRIPTION
**Pull request by a-rodian-jedi**
#### Brief overview of PR changes/additions

- Fixes a memory leak in Portal Telnet connections
- Forces weak references on Telnet negotiation objects
- Stops _nop_keepalive LoopingCall on disconnected sessions

#### Motivation for adding to Evennia

- Community complaints about memory leaks

#### Other info (issues closed, discussion etc)

- #3225 
- This is not everything. There is also a leak in AttributeHandler, I believe, but I will separate that. I think these fixes should come 1 at a time for stability.

### CODE REVIEW

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:1]: I know this is meant as a WIP, but if it should go into Evennia core, this requires a useful and descriptive doc string including usage examples and maybe even links to further reading - specially since the contents is so technical.

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:46]: Leftover code?

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:48]: Docstring on Google code style needed. See CODING_STYLE.txt

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:91]: Docstrings are required for all functions and methods. We don't use Python typehints at this time, so adhering to Google codestyle type descriptors is paramount.

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:114]: All functions need docstrings.

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:137]: Docstring needed.

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:153]: Docstring needed. See CODING_STYLE.txt

**Review by Griatch — CHANGES_REQUESTED** [review]
Thanks a lot for your time spent working on this! :D  I'd like to better understand the workings of the solution though. For profiler, more docs are needed.

File: evennia/server/portal/mccp.py
Comment (a-rodian-jedi) [reply evennia/server/portal/mccp.py:61]: Hello there! This `()` is actually how Python dereferences a pointer. See this code here for `__call__` in weakref: https://github.com/python/cpython/blob/3.11/Lib/weakref.py#L68C14-L68C14

File: evennia/server/profiling/profiler.py
Comment (a-rodian-jedi) [reply evennia/server/profiling/profiler.py:46]: I didn't get the function wrapper to ever get the right frame. I think I could fix this now, but at least for the next week or so my priorities will be elsewhere. I can just remove this for now and add it later.

**Comment by Griatch** [comment]
@a-rodian-jedi When you are ready for a re-review, you can note it here or click the little spinner up in the reviewer list :)

File: evennia/server/portal/mccp.py
Comment (Griatch) [reply evennia/server/portal/mccp.py:61]: Interesting! Something new learned every day.

File: evennia/server/profiling/profiler.py
Comment (Griatch) [inline evennia/server/profiling/profiler.py:11]: This is a good doc string, but it's missing a usage example of how to actually _use_ this tool - that is, in which situation - from inside code, from the command line, how to generate the needed tracemalloc snapshots etc. This is because people _will_ find this code and ask about how to use it otherwise ;)

**Review by Griatch — CHANGES_REQUESTED** [review]
Thanks for the additional documentation. This looks good. We just need one additional bit of info - a usage example of how to use the memory profiler. Since this is going into core, people are bound to find this code and want to use it, so they should have the needed info to at least test it out (rather than having to go into discord/discussion thread).

**Comment by Griatch** [comment]
@a-rodian-jedi Ping; apart from the PR feedback, this also needs to be rebased against the upstream repo (that's why there are so many extra 'changed files'; this is not your fault, but because the upstream repo changed to make it a smaller download. Sorry for the inconvenience!

**Comment by Griatch** [comment]
@a-rodian-jedi Reminder; this is waiting for some fixes before it can merge.

**Comment by Griatch** [comment]
@a-rodian-jedi It would be good to get this one wrapped up, it's so little left to do on it. :)

**Comment by Griatch** [comment]
Ping. Just a usage example missing!

**Comment by Griatch** [comment]
@a-rodian-jedi Another ping :)

**Comment by Griatch** [comment]
@a-rodian-jedi Ping :)

**Pull request closed without merging**
